File: mobile/genie_ai_mobile_swiftui/GenieAI/Services/ConfigService.swift
// ConfigService.swift
// Loads and provides access to genie-ai-config.json

import Foundation

struct AppConfig: Codable {
    let app: AppInfo
    let theme: ThemeConfig
    let features: FeaturesConfig?
    let custom: [String: AnyCodable]?

    struct AppInfo: Codable {
        let title: String
        let icon: IconConfig

        struct IconConfig: Codable {
            let type: String
            let value: String
        }
    }

    struct ThemeConfig: Codable {
        let primaryColor: String
        let secondaryColor: String
        let backgroundColor: String?
        let textColor: String?
        let navbar: NavbarConfig

        struct NavbarConfig: Codable {
            let gradientStart: String
            let gradientEnd: String
            let textColor: String
        }
    }

    struct FeaturesConfig: Codable {
        let chat: ChatConfig?

        struct ChatConfig: Codable {
            let welcomeMessage: String?
            let botName: String?
            let quickHelp: QuickHelpConfig?

            struct QuickHelpConfig: Codable {
                let layout: LayoutConfig?
                let defaults: DefaultsConfig?
                let buttons: [QuickHelpButtonConfig]?

                struct LayoutConfig: Codable {
                    let columns: Int?
                    let gapX: String?
                    let gapY: String?
                    let childAspectRatio: Double?
                }

                struct DefaultsConfig: Codable {
                    let height: String?
                    let fontSize: String?
                    let borderRadius: String?
                    let showShadow: Bool?
                    let showBorder: Bool?
                }

                struct QuickHelpButtonConfig: Codable {
                    let id: String
                    let category: String?
                    let action: ActionConfig
                    let appearance: AppearanceConfig?

                    struct ActionConfig: Codable {
                        let visibleText: String
                        let hiddenPrompt: String
                    }

                    struct AppearanceConfig: Codable {
                        let label: LabelConfig?
                        let icon: IconConfig?
                        let style: StyleConfig?
                        let darkMode: DarkModeConfig?

                        struct LabelConfig: Codable {
                            let text: String?
                            let color: String?
                        }

                        struct IconConfig: Codable {
                            let value: String?
                            let color: String?
                        }

                        struct StyleConfig: Codable {
                            let background: BackgroundConfig?

                            struct BackgroundConfig: Codable {
                                let gradient: GradientConfig?

                                struct GradientConfig: Codable {
                                    let start: String?
                                    let end: String?
                                    let direction: String?
                                }
                            }
                        }

                        struct DarkModeConfig: Codable {
                            let label: LabelConfig?
                            let icon: IconConfig?
                            let style: StyleConfig?
                        }
                    }
                }
            }
        }
    }
}

// Helper for encoding/decoding arbitrary JSON
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let intValue = try? container.decode(Int.self) {
            value = intValue
        } else if let doubleValue = try? container.decode(Double.self) {
            value = doubleValue
        } else if let stringValue = try? container.decode(String.self) {
            value = stringValue
        } else if let boolValue = try? container.decode(Bool.self) {
            value = boolValue
        } else if let arrayValue = try? container.decode([AnyCodable].self) {
            value = arrayValue.map { $0.value }
        } else if let dictValue = try? container.decode([String: AnyCodable].self) {
            value = dictValue.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch value {
        case let intValue as Int:
            try container.encode(intValue)
        case let doubleValue as Double:
            try container.encode(doubleValue)
        case let stringValue as String:
            try container.encode(stringValue)
        case let boolValue as Bool:
            try container.encode(boolValue)
        case let arrayValue as [Any]:
            try container.encode(arrayValue.map { AnyCodable($0) })
        case let dictValue as [String: Any]:
            try container.encode(dictValue.mapValues { AnyCodable($0) })
        default:
            try container.encodeNil()
        }
    }
}

@Observable
class ConfigService {
    static let shared = ConfigService()

    private(set) var config: AppConfig?
    private(set) var isLoaded = false

    var appTitle: String {
        config?.app.title ?? "Genie AI"
    }

    var welcomeMessage: String {
        config?.features?.chat?.welcomeMessage ?? "Welcome to Genie AI!"
    }

    var botName: String {
        config?.features?.chat?.botName ?? "Genie AI"
    }

    var primaryColor: String {
        config?.theme.primaryColor ?? "#4682B4"
    }

    var secondaryColor: String {
        config?.theme.secondaryColor ?? "#5F9EA0"
    }

    var backgroundColor: String {
        config?.theme.backgroundColor ?? "#D3E0EA"
    }

    var textColor: String {
        config?.theme.textColor ?? "#1C2526"
    }

    var navbarGradientStart: String {
        config?.theme.navbar.gradientStart ?? "#4682B4"
    }

    var navbarGradientEnd: String {
        config?.theme.navbar.gradientEnd ?? "#5F9EA0"
    }

    var navbarTextColor: String {
        config?.theme.navbar.textColor ?? "#F0F8FF"
    }

    var quickHelpButtons: [AppConfig.FeaturesConfig.ChatConfig.QuickHelpConfig.QuickHelpButtonConfig] {
        config?.features?.chat?.quickHelp?.buttons ?? []
    }

    private init() {
        loadConfig()
    }

    private func loadConfig() {
        guard let url = Bundle.main.url(forResource: "genie-ai-config", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            print("[ConfigService] Failed to load config file")
            return
        }

        do {
            config = try JSONDecoder().decode(AppConfig.self, from: data)
            isLoaded = true
            print("[ConfigService] Config loaded successfully")
        } catch {
            print("[ConfigService] Failed to decode config: \(error)")
        }
    }
}

File: mobile/genie_ai_mobile_compose/app/src/main/java/com/genieai/mobile/service/ConfigService.kt
package com.genieai.mobile.service

import android.content.Context
import android.util.Log
import androidx.compose.ui.graphics.Color
import com.genieai.mobile.GenieAIApplication
import com.genieai.mobile.data.model.QuickHelpButton
import com.google.gson.Gson
import com.google.gson.JsonObject

object ConfigService {

    private const val TAG = "ConfigService"
    private var config: JsonObject? = null
    private val gson = Gson()

    val appTitle: String
        get() = config?.getAsJsonObject("app")?.get("title")?.asString ?: "Genie AI"

    val primaryColor: String
        get() = config?.getAsJsonObject("theme")?.get("primaryColor")?.asString ?: "#4682B4"

    val secondaryColor: String
        get() = config?.getAsJsonObject("theme")?.get("secondaryColor")?.asString ?: "#5F9EA0"

    val backgroundColor: String
        get() = config?.getAsJsonObject("theme")?.get("backgroundColor")?.asString ?: "#D3E0EA"

    val textColor: String
        get() = config?.getAsJsonObject("theme")?.get("textColor")?.asString ?: "#1C2526"

    val navbarGradientStart: String
        get() = config?.getAsJsonObject("theme")?.getAsJsonObject("navbar")
            ?.get("gradientStart")?.asString ?: "#4682B4"

    val navbarGradientEnd: String
        get() = config?.getAsJsonObject("theme")?.getAsJsonObject("navbar")
            ?.get("gradientEnd")?.asString ?: "#5F9EA0"

    val navbarTextColor: String
        get() = config?.getAsJsonObject("theme")?.getAsJsonObject("navbar")
            ?.get("textColor")?.asString ?: "#F0F8FF"

    val welcomeMessage: String
        get() = config?.getAsJsonObject("features")?.getAsJsonObject("chat")
            ?.get("welcomeMessage")?.asString ?: "Welcome to Genie AI"

    val botName: String
        get() = config?.getAsJsonObject("features")?.getAsJsonObject("chat")
            ?.get("botName")?.asString ?: "Genie AI"

    fun init(context: Context = GenieAIApplication.instance) {
        try {
            val inputStream = context.resources.openRawResource(
                context.resources.getIdentifier("genie_ai_config", "raw", context.packageName)
            )
            val jsonString = inputStream.bufferedReader().use { it.readText() }
            config = gson.fromJson(jsonString, JsonObject::class.java)
            Log.d(TAG, "Config loaded: ${config?.get("app")}")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load config: ${e.message}", e)
        }
    }

    fun getQuickHelpButtons(): List<QuickHelpButton> {
        val buttons = mutableListOf<QuickHelpButton>()
        val quickHelp = config?.getAsJsonObject("features")
            ?.getAsJsonObject("chat")
            ?.getAsJsonObject("quickHelp") ?: return buttons

        val buttonsArray = quickHelp.getAsJsonArray("buttons") ?: return buttons

        for (element in buttonsArray) {
            val btn = element.asJsonObject
            val appearance = btn.getAsJsonObject("appearance") ?: continue
            val action = btn.getAsJsonObject("action") ?: continue
            val style = appearance.getAsJsonObject("style")
            val darkMode = appearance.getAsJsonObject("darkMode")
            val gradient = style?.getAsJsonObject("background")?.getAsJsonObject("gradient")
            val darkStyle = darkMode?.getAsJsonObject("style")
            val darkGradient = darkStyle?.getAsJsonObject("background")?.getAsJsonObject("gradient")

            val rawLabel = appearance.getAsJsonObject("label")?.get("text")?.asString ?: ""
            val rawVisible = action.get("visibleText")?.asString ?: ""
            val rawHidden = action.get("hiddenPrompt")?.asString ?: ""

            buttons.add(
                QuickHelpButton(
                    id = btn.get("id")?.asString ?: "",
                    category = btn.get("category")?.let { if (it.isJsonNull) null else it.asString },
                    labelText = resolveI18nKey(rawLabel),
                    labelColor = parseColor(appearance.getAsJsonObject("label")?.get("color")?.asString),
                    iconPath = appearance.getAsJsonObject("icon")?.get("value")?.asString ?: "",
                    iconColor = parseColor(appearance.getAsJsonObject("icon")?.get("color")?.asString),
                    gradientStart = parseColor(gradient?.get("start")?.asString),
                    gradientEnd = parseColor(gradient?.get("end")?.asString),
                    darkLabelColor = parseColor(darkMode?.getAsJsonObject("label")?.get("color")?.asString),
                    darkIconColor = parseColor(darkMode?.getAsJsonObject("icon")?.get("color")?.asString),
                    darkGradientStart = parseColor(darkGradient?.get("start")?.asString),
                    darkGradientEnd = parseColor(darkGradient?.get("end")?.asString),
                    visibleText = resolveI18nKey(rawVisible),
                    hiddenPrompt = resolveI18nKey(rawHidden)
                )
            )
        }
        return buttons
    }

    fun getQuickHelpColumns(): Int {
        return config?.getAsJsonObject("features")
            ?.getAsJsonObject("chat")
            ?.getAsJsonObject("quickHelp")
            ?.getAsJsonObject("layout")
            ?.get("columns")?.asInt ?: 2
    }

    /**
     * Resolve an i18n key (e.g. "quickhelp.justChat") to its localized string.
     * Converts dotted keys to Android resource names (quickhelp.justChat → quickhelp_just_chat),
     * then looks up via resources. Falls back to the raw key if not found.
     */
    fun resolveI18nKey(key: String): String {
        if (key.isBlank()) return key
        val context = GenieAIApplication.instance
        // Convert dotted camelCase key to snake_case resource name
        // e.g. "quickhelp.justChat" → "quickhelp_just_chat"
        // e.g. "quickhelp.applyForIDUserPrompt" → "quickhelp_apply_for_id_user_prompt"
        val resourceName = key
            .replace(".", "_")
            .replace(Regex("([a-z])([A-Z])")) { "${it.groupValues[1]}_${it.groupValues[2]}" }
            .lowercase()
        val resId = context.resources.getIdentifier(resourceName, "string", context.packageName)
        return if (resId != 0) context.getString(resId) else key
    }

    private fun parseColor(hex: String?): Color {
        if (hex == null) return Color.Unspecified
        return try {
            Color(android.graphics.Color.parseColor(hex))
        } catch (_: Exception) {
            Color.Unspecified
        }
    }
}
